src/vmm/src/lib.rs

```diff
@@ -0,0 +1,404 @@
+        bootparams.hdr.cmdline_size = kernel_cfg.cmdline.len() as u32 + 1;
+
+        // Load the kernel command line into guest memory.
+        let mut cmdline = Cmdline::new(kernel_cfg.cmdline.len() + 1);
+        cmdline
+            .insert_str(kernel_cfg.cmdline)
+            .map_err(Error::Cmdline)?;
+        load_cmdline(
+            &self.guest_memory,
+            GuestAddress(CMDLINE_START),
+            // Safe because the command line is valid.
+            unsafe { &CString::from_vec_unchecked(cmdline.into()) },
+        )
+        .map_err(Error::KernelLoad)?;
+
+        // Write the boot parameters in the zeropage.
+        LinuxBootConfigurator::write_bootparams::<GuestMemoryMmap>(
+            &BootParams::new::<boot_params>(&bootparams, zero_page_addr),
+            &self.guest_memory,
+        )
+        .map_err(Error::BootConfigure)?;
+
+        Ok(kernel_load)
+    }
+
+    /// Configure PIO devices.
+    ///
+    /// This sets up the following PIO devices:
+    /// * `x86_64`: serial console
+    /// * `aarch64`: N/A
+    pub fn configure_pio_devices(&mut self) -> Result<()> {
+        // First, create the irqchip.
+        // On `x86_64`, this _must_ be created _before_ the vCPUs.
+        // It sets up the virtual IOAPIC, virtual PIC, and sets up the future vCPUs for local APIC.
+        // When in doubt, look in the kernel for `KVM_CREATE_IRQCHIP`.
+        // https://elixir.bootlin.com/linux/latest/source/arch/x86/kvm/x86.c
+        self.vm_fd.create_irq_chip().map_err(Error::KvmIoctl)?;
+
+        // Set up the speaker PIT, because some kernels are musical and access the speaker port
+        // during boot. Without this, KVM would continuously exit to userspace.
+        // TODO: does this *need* the irqchip in place?
+        let mut pit_config = kvm_pit_config::default();
+        pit_config.flags = KVM_PIT_SPEAKER_DUMMY;
+        self.vm_fd
+            .create_pit2(pit_config)
+            .map_err(Error::KvmIoctl)?;
+
+        // Create the serial console.
+        let interrupt_evt = EventFd::new(libc::EFD_NONBLOCK).map_err(Error::IO)?;
+        let serial = Arc::new(Mutex::new(SerialWrapper(Serial::new(
+            interrupt_evt.try_clone().map_err(Error::IO)?,
+            stdout(),
+        ))));
+        // Register its interrupt fd with KVM.
+        self.vm_fd
+            .register_irqfd(&interrupt_evt, 4)
+            .map_err(Error::KvmIoctl)?;
+
+        // Configure event management for the serial console's events.
+        self.device_mgr.serial_ev_mgr.add_subscriber(serial.clone());
+        self.device_mgr.serial = Some(serial);
+
+        Ok(())
+    }
+
+    /// Configure guest vCPUs.
+    ///
+    /// # Arguments
+    ///
+    /// * `vcpu_cfg` - [`VcpuConfig`](struct.VcpuConfig.html) struct containing vCPU configurations.
+    pub fn configure_vcpus(
+        &mut self,
+        vcpu_cfg: VcpuConfig,
+        kernel_load: KernelLoaderResult,
```

Comment:
The arguments list should be updated with this one. Also, isn't  the `kernel_load` GuestAddress field from `KernelLoaderResult` enough here or do we expect to use in the future the other fields too?